ofproto/ofproto.c

```diff
@@ -5552,6 +5552,75 @@ handle_load_filter_prog(struct ofconn *ofconn, const struct ofp_header *oh)
     return error;
 }
 
+static enum ofperr
+handle_update_map(struct ofconn *ofconn, const struct ofp_header *oh)
+    OVS_EXCLUDED(ofproto_mutex)
+{
+    struct ofproto *ofproto = ofconn_get_ofproto(ofconn);
+    enum ofperr error;
+    error = reject_slave_controller(ofconn);
+    if (error) {
+        return error;
+    }
+
+    struct ol_update_map msg;
+    void *data = NULL;
+    error = ofputil_decode_update_map(&msg, &data, oh);
+    if (error) {
+        return error;
+    }
+
+    struct ubpf_vm *vm = ubpf_vms_lookup(ofproto, msg.filter_prog);
+    if (!vm) {
+        VLOG_WARN_RL(&rl,
+                     "The referenced filter program could not be found.");
+        return OFPERR_OFPBRC_EPERM;
+    }
+
+
+    if (msg.map_id >= vm->nb_maps) {
+        VLOG_WARN_RL(&rl,
+                     "The map with provided identifier does not exist.");
+        return OFPERR_OFPBRC_EPERM;
+    }
+
+    const char *name = vm->ext_map_names[msg.map_id];
+    if(!name) {
+        VLOG_WARN_RL(&rl,
+                     "The map with the given ID does not exist.");
+        return OFPERR_OFPBRC_EPERM;
+    }
+
+    struct ubpf_map *map = ubpf_lookup_registered_map(vm, name);
+
+    if(msg.key_size != map->key_size) {
+        VLOG_WARN_RL(&rl,
+                     "Size of key does not match to size of map's key");
+        return OFPERR_OFPBRC_EPERM;
+    }
+
+    if(msg.value_size != map->value_size) {
+        VLOG_WARN_RL(&rl,
+                     "Size of value does not match to size of map's value");
+        return OFPERR_OFPBRC_EPERM;
+    }
+
+    for(int i = 0; i < msg.nb_elems; i++) {
+        void *key = data;
+        data += msg.key_size;
+        void *value = data;
+        data += msg.value_size;
+        if (ubpf_map_update(map, key, value) < 0) {
+            // TODO: not sure how to return failed one in the batch
```

Comment:
I think failing the whole batch is the appropriate solution here, for now at least. You can remove the TODO.